packages/data/src/types.d.ts

```diff
@@ -0,0 +1,36 @@
+export type WPDataFunctionOrGeneratorArray = {
+    [index: number]: Function|Generator;
+};
+export type WPDataFunctionArray = {
+    [index: number]: Function;
+};
+
+export type WPDataAttachedStore = {
+    getSelectors: () => WPDataFunctionArray,
+    getActions: () => WPDataFunctionArray,
+    subscribe: (listener: () => void) => (() => void)
+};
+
+export type WPDataStoreDefinition = {
+    /**
+     * Store Name
+     */
+    name: string,
+
+    /**
+     * Store configuration object.
+     */
+    __internalAttach: (registry: WPDataRegistry) => WPDataAttachedStore,
+};
+
+export type WPDataReduxStoreConfig = {
+    reducer: ( state: any, action: any ) => any,
+    actions?: WPDataFunctionOrGeneratorArray,
+    resolvers?: WPDataFunctionOrGeneratorArray,
+    selectors?: WPDataFunctionArray,
+    controls?: WPDataFunctionArray,
+}
+
+export type WPDataRegistry = {
```

Comment:
@sirreal I added some unperfect types to the data module without actually checking them for now. I'm not sure this is a good idea but for now, I can type everything (not even the file I wanted to type redux-store) properly because of the dependencies.

Is this fine or should I remove all the types for now?

Reply:
I'd prefer not to add manual type declarations, so I'd discourage committing this file. I can help to translate these to "live" types living in the source but you may not want to wait on me. In that case, I'd recommend removing.

---

(I think there's a typo: you _cannot_ type the data package. I'm replying with that assumption 🙂 )

There's precedent for typing parts of a package. We're working on `components` right now. We can start progressively typing like we've done there if you'd like to get `redux-store` typed:

https://github.com/WordPress/gutenberg/blob/e5c3c66326d23ca20c6c8e183eb31a7735ee6703/packages/components/tsconfig.json#L13-L19

I'm happy to prep the data package for progressive typing if you'd like.

There are other options like relaxing some rules, but I'd like to start by progressively typing things.


---

This approach does make me very curious. I've been tempted to do something similar in the past because there are type-level things that cannot be expressed in JSDoc syntax. At the very least, JSDoc is much more cumbersome than TypeScript (in my experience) to express the same thing.

While I discourage committing the declaration file, there's some appeal to committing it. If we start committing declaration files, I'm curious where it would lead. Declaration files are effectively TypeScript minus the JavaScript 🤔 

---

Aside: `compose` has been a very challenging package to type but we're getting more interest and expertise and may be able to tackle it soon, unblocking types for the data package.

Reply:
So my question was not about the declaration type file. I actually  think it's  a good idea to commit it and use `import` to reference it instead of adding typedef comments on the top of JS files (This is also shown on the typescript JSDoc documentation).

I was more asking about the second part. I tried typing `redux-store/index.js` file entirely and use the same technique used for the components package to only check this file but even this was not sufficient enough because that file has 2/3 untyped dependencies and I didn't want to go into a rabbit whole, so I was wondering if I should leave the types file even if it's currently unchecked.

Reply:
> I was more asking about the second part. I tried typing redux-store/index.js file entirely and use the same technique used for the components package to only check this file but even this was not sufficient enough because that file has 2/3 untyped dependencies and I didn't want to go into a rabbit whole, so I was wondering if I should leave the types file even if it's currently unchecked.

Yes, it's very difficult to jump into the middle of a package and type just a little bit. It can be made easier by liberally adding `// @ts-ignore` or be relaxing the rules. Setting the the compiler option `noImplicitAny` to `false` can make things easier but sacrifices safety. I've created a PR (based off this branch) that explores some of this if it's helpful: https://github.com/WordPress/gutenberg/pull/27030

I had assumed the declaration file was declaring types for a module that exists but it is not. In this case it's describing interfaces that are used throughout which seems like an excellent choice. Following this pattern will make some of our other typing work easier 🙂